// @neo4j/graphql/dist/schema/validation/directives.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.writeonlyDirective = exports.timestampDirective = exports.relationshipDirective = exports.readonlyDirective = exports.privateDirective = exports.ignoreDirective = exports.idDirective = exports.excludeDirective = exports.defaultDirective = exports.cypherDirective = exports.coalesceDirective = void 0;
const graphql_1 = require("graphql");
const enums_1 = require("./enums");
const scalars_1 = require("./scalars");
exports.coalesceDirective = new graphql_1.GraphQLDirective({
    name: "coalesce",
    description: "Instructs @neo4j/graphql to wrap the property in a coalesce() function during queries, using the single value specified.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
    args: {
        value: {
            description: "The value to use in the coalesce() function. Must be a scalar type and must match the type of the field with which this directive decorates.",
            type: new graphql_1.GraphQLNonNull(scalars_1.ScalarType),
        },
    },
});
exports.cypherDirective = new graphql_1.GraphQLDirective({
    name: "cypher",
    description: "Instructs @neo4j/graphql to run the specified Cypher statement in order to resolve the value of the field to which the directive is applied.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
    args: {
        statement: {
            description: "The Cypher statement to run which returns a value of the same type composition as the field definition on which the directive is applied.",
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
        },
    },
});
exports.defaultDirective = new graphql_1.GraphQLDirective({
    name: "default",
    description: "Instructs @neo4j/graphql to set the specified value as the default value in the CreateInput type for the object type in which this directive is used.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
    args: {
        value: {
            description: "The default value to use. Must be a scalar type and must match the type of the field with which this directive decorates.",
            type: new graphql_1.GraphQLNonNull(scalars_1.ScalarType),
        },
    },
});
exports.excludeDirective = new graphql_1.GraphQLDirective({
    name: "exclude",
    description: "Instructs @neo4j/graphql to exclude the specified operations from query and mutation generation. If used without an argument, no queries or mutations will be generated for this type.",
    locations: [graphql_1.DirectiveLocation.OBJECT],
    args: {
        operations: {
            defaultValue: enums_1.ExcludeOperationEnum.getValues().map((v) => v.value),
            type: new graphql_1.GraphQLNonNull(new graphql_1.GraphQLList(new graphql_1.GraphQLNonNull(enums_1.ExcludeOperationEnum))),
        },
    },
});
exports.idDirective = new graphql_1.GraphQLDirective({
    name: "id",
    description: "Indicates that the field is the unique identifier for the object type, and additionally enables the autogeneration of IDs.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
    args: {
        autogenerate: {
            defaultValue: false,
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLBoolean),
        },
    },
});
exports.ignoreDirective = new graphql_1.GraphQLDirective({
    name: "ignore",
    description: "Instructs @neo4j/graphql to completely ignore a field definition, assuming that it will be fully accounted for by custom resolvers.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
});
exports.privateDirective = new graphql_1.GraphQLDirective({
    name: "private",
    description: "Instructs @neo4j/graphql to only expose a field through the Neo4j GraphQL OGM.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
});
exports.readonlyDirective = new graphql_1.GraphQLDirective({
    name: "readonly",
    description: "Instructs @neo4j/graphql to only include a field in generated input type for creating, and in the object type within which the directive is applied.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
});
exports.relationshipDirective = new graphql_1.GraphQLDirective({
    name: "relationship",
    description: "Instructs @neo4j/graphql to treat this field as a relationship. Opens up the ability to create and connect on this field.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
    args: {
        type: {
            type: new graphql_1.GraphQLNonNull(graphql_1.GraphQLString),
        },
        direction: {
            type: new graphql_1.GraphQLNonNull(enums_1.RelationshipDirectionEnum),
        },
    },
});
exports.timestampDirective = new graphql_1.GraphQLDirective({
    name: "timestamp",
    description: "Instructs @neo4j/graphql to generate timestamps on particular events, which will be available as the value of the specified field.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
    args: {
        operations: {
            description: "Which events to generate timestamps on. Defaults to both create and update.",
            defaultValue: enums_1.TimestampOperationEnum.getValues().map((v) => v.value),
            type: new graphql_1.GraphQLNonNull(new graphql_1.GraphQLList(new graphql_1.GraphQLNonNull(enums_1.TimestampOperationEnum))),
        },
    },
});
exports.writeonlyDirective = new graphql_1.GraphQLDirective({
    name: "writeonly",
    description: "Instructs @neo4j/graphql to only include a field in the generated input types for the object type within which the directive is applied, but exclude it from the object type itself.",
    locations: [graphql_1.DirectiveLocation.FIELD_DEFINITION],
});
