// @neo4j/graphql/dist/translate/create-where-and-params.test.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const create_where_and_params_1 = __importDefault(require("./create-where-and-params"));
describe("createWhereAndParams", () => {
    test("should be a function", () => {
        expect(create_where_and_params_1.default).toBeInstanceOf(Function);
    });
    test("should return the correct clause with 1 param", () => {
        const whereInput = {
            title: "some title",
        };
        const varName = "this";
        // @ts-ignore
        const node = {
            relationFields: [],
            cypherFields: [],
            enumFields: [],
            scalarFields: [],
            primitiveFields: [],
            dateTimeFields: [],
            interfaceFields: [],
            unionFields: [],
            objectFields: [],
            pointFields: [],
            otherDirectives: [],
            interfaces: [],
        };
        // @ts-ignore
        const context = { neoSchema: { nodes: [] } };
        const result = create_where_and_params_1.default({ whereInput, varName, node, context });
        expect(result[0]).toEqual(`WHERE this.title = $this_title`);
        expect(result[1]).toMatchObject({ this_title: whereInput.title });
    });
});
