// @neo4j/graphql/dist/translate/create-auth-and-params.test.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const randomstring_1 = require("randomstring");
const create_auth_and_params_1 = __importDefault(require("./create-auth-and-params"));
const utils_1 = require("../utils");
describe("createAuthAndParams", () => {
    describe("operations", () => {
        test("should cover all rules when using operations *", () => {
            const idField = {
                fieldName: "id",
                typeMeta: {
                    name: "ID",
                    array: false,
                    required: false,
                    pretty: "String",
                    input: {
                        where: {
                            type: "String",
                            pretty: "String",
                        },
                        create: {
                            type: "String",
                            pretty: "String",
                        },
                        update: {
                            type: "String",
                            pretty: "String",
                        },
                    },
                },
                otherDirectives: [],
                arguments: [],
            };
            // @ts-ignore
            const node = {
                name: "Movie",
                relationFields: [],
                cypherFields: [],
                enumFields: [],
                scalarFields: [],
                primitiveFields: [idField],
                dateTimeFields: [],
                interfaceFields: [],
                objectFields: [],
                pointFields: [],
                authableFields: [idField],
                auth: {
                    rules: [
                        { allow: { id: "$jwt.sub" } },
                        { operations: ["CREATE"], roles: ["admin"] },
                        { roles: ["admin"] },
                    ],
                    type: "JWT",
                },
            };
            // @ts-ignore
            const neoSchema = {
                nodes: [node],
            };
            const sub = randomstring_1.generate({
                charset: "alphabetic",
            });
            // @ts-ignore
            const context = { neoSchema, jwt: { sub } };
            const result = create_auth_and_params_1.default({
                context,
                entity: node,
                operation: "READ",
                allow: { parentNode: node, varName: "this" },
            });
            expect(utils_1.trimmer(result[0])).toEqual(utils_1.trimmer(`
                    EXISTS(this.id) AND this.id = $this_auth_allow0_id OR ANY(r IN ["admin"] WHERE ANY(rr IN $auth.roles WHERE r = rr))
                `));
            expect(result[1]).toMatchObject({
                this_auth_allow0_id: sub,
            });
        });
    });
    describe("rules", () => {
        test("should showcase the default OR behavior of stacked rules", () => {
            const idField = {
                fieldName: "id",
                typeMeta: {
                    name: "ID",
                    array: false,
                    required: false,
                    pretty: "String",
                    input: {
                        where: {
                            type: "String",
                            pretty: "String",
                        },
                        create: {
                            type: "String",
                            pretty: "String",
                        },
                        update: {
                            type: "String",
                            pretty: "String",
                        },
                    },
                },
                otherDirectives: [],
                arguments: [],
            };
            // @ts-ignore
            const node = {
                name: "Movie",
                relationFields: [],
                cypherFields: [],
                enumFields: [],
                scalarFields: [],
                primitiveFields: [idField],
                dateTimeFields: [],
                interfaceFields: [],
                objectFields: [],
                pointFields: [],
                authableFields: [idField],
                auth: {
                    rules: [{ allow: { id: "$jwt.sub" } }, { roles: ["admin"] }],
                    type: "JWT",
                },
            };
            // @ts-ignore
            const neoSchema = {
                nodes: [node],
            };
            const sub = randomstring_1.generate({
                charset: "alphabetic",
            });
            // @ts-ignore
            const context = { neoSchema };
            context.jwt = {
                sub,
            };
            const result = create_auth_and_params_1.default({
                context,
                entity: node,
                allow: { parentNode: node, varName: "this" },
            });
            expect(utils_1.trimmer(result[0])).toEqual(utils_1.trimmer(`
                    EXISTS(this.id) AND this.id = $this_auth_allow0_id OR ANY(r IN ["admin"] WHERE ANY(rr IN $auth.roles WHERE r = rr))
                `));
            expect(result[1]).toMatchObject({
                this_auth_allow0_id: sub,
            });
        });
        test("should showcase the default AND default behavior of the keys in the rule", () => {
            const idField = {
                fieldName: "id",
                typeMeta: {
                    name: "ID",
                    array: false,
                    required: false,
                    pretty: "String",
                    input: {
                        where: {
                            type: "String",
                            pretty: "String",
                        },
                        create: {
                            type: "String",
                            pretty: "String",
                        },
                        update: {
                            type: "String",
                            pretty: "String",
                        },
                    },
                },
                otherDirectives: [],
                arguments: [],
            };
            // @ts-ignore
            const node = {
                name: "Movie",
                relationFields: [],
                cypherFields: [],
                enumFields: [],
                scalarFields: [],
                primitiveFields: [idField],
                dateTimeFields: [],
                interfaceFields: [],
                objectFields: [],
                pointFields: [],
                authableFields: [idField],
                auth: {
                    rules: [{ allow: { id: "$jwt.sub" }, roles: ["admin"] }],
                    type: "JWT",
                },
            };
            // @ts-ignore
            const neoSchema = {
                nodes: [node],
            };
            const sub = randomstring_1.generate({
                charset: "alphabetic",
            });
            // @ts-ignore
            const context = { neoSchema };
            context.jwt = {
                sub,
            };
            const result = create_auth_and_params_1.default({
                context,
                entity: node,
                allow: { parentNode: node, varName: "this" },
            });
            expect(utils_1.trimmer(result[0])).toEqual(utils_1.trimmer(`
                     ANY(r IN ["admin"] WHERE ANY(rr IN $auth.roles WHERE r = rr)) AND EXISTS(this.id) AND this.id = $this_auth_allow0_id
                `));
            expect(result[1]).toMatchObject({
                this_auth_allow0_id: sub,
            });
        });
    });
    describe("top-level keys", () => {
        test("AND OR", () => {
            ["AND", "OR"].forEach((key) => {
                const idField = {
                    fieldName: "id",
                    typeMeta: {
                        name: "ID",
                        array: false,
                        required: false,
                        pretty: "String",
                        input: {
                            where: {
                                type: "String",
                                pretty: "String",
                            },
                            create: {
                                type: "String",
                                pretty: "String",
                            },
                            update: {
                                type: "String",
                                pretty: "String",
                            },
                        },
                    },
                    otherDirectives: [],
                    arguments: [],
                };
                // @ts-ignore
                const node = {
                    name: "Movie",
                    relationFields: [],
                    cypherFields: [],
                    enumFields: [],
                    scalarFields: [],
                    primitiveFields: [idField],
                    dateTimeFields: [],
                    interfaceFields: [],
                    objectFields: [],
                    pointFields: [],
                    authableFields: [idField],
                    auth: {
                        rules: [{ [key]: [{ allow: { id: "$jwt.sub" } }, { roles: ["admin"] }] }],
                        type: "JWT",
                    },
                };
                // @ts-ignore
                const neoSchema = {
                    nodes: [node],
                };
                const sub = randomstring_1.generate({
                    charset: "alphabetic",
                });
                // @ts-ignore
                const context = { neoSchema };
                context.jwt = {
                    sub,
                };
                const result = create_auth_and_params_1.default({
                    context,
                    entity: node,
                    allow: { parentNode: node, varName: "this" },
                });
                expect(utils_1.trimmer(result[0])).toEqual(utils_1.trimmer(`
                        EXISTS(this.id) AND this.id = $this${key}0_auth_allow0_id ${key} ANY(r IN ["admin"] WHERE ANY(rr IN $auth.roles WHERE r = rr))
                    `));
                expect(result[1]).toMatchObject({
                    [`this${key}0_auth_allow0_id`]: sub,
                });
            });
        });
        test("all keys used together", () => {
            const idField = {
                fieldName: "id",
                typeMeta: {
                    name: "ID",
                    array: false,
                    required: false,
                    pretty: "String",
                    input: {
                        where: {
                            type: "String",
                            pretty: "String",
                        },
                        create: {
                            type: "String",
                            pretty: "String",
                        },
                        update: {
                            type: "String",
                            pretty: "String",
                        },
                    },
                },
                otherDirectives: [],
                arguments: [],
            };
            // @ts-ignore
            const node = {
                name: "Movie",
                relationFields: [],
                cypherFields: [],
                enumFields: [],
                scalarFields: [],
                primitiveFields: [idField],
                dateTimeFields: [],
                interfaceFields: [],
                objectFields: [],
                pointFields: [],
                authableFields: [idField],
                auth: {
                    rules: [
                        {
                            roles: ["admin"],
                            allow: { id: "$jwt.sub" },
                            AND: [{ allow: { id: "$jwt.sub" } }, { roles: ["admin"] }],
                            OR: [{ allow: { id: "$jwt.sub" } }, { roles: ["admin"] }],
                        },
                    ],
                    type: "JWT",
                },
            };
            // @ts-ignore
            const neoSchema = {
                nodes: [node],
            };
            const sub = randomstring_1.generate({
                charset: "alphabetic",
            });
            // @ts-ignore
            const context = { neoSchema };
            context.jwt = {
                sub,
            };
            const result = create_auth_and_params_1.default({
                context,
                entity: node,
                allow: { parentNode: node, varName: "this" },
            });
            expect(utils_1.trimmer(result[0])).toEqual(utils_1.trimmer(`
                    ANY(r IN ["admin"] WHERE ANY(rr IN $auth.roles WHERE r = rr))
                AND
                    EXISTS(this.id) AND this.id = $this_auth_allow0_id
                AND
                    EXISTS(this.id) AND this.id = $thisAND0_auth_allow0_id AND ANY(r IN ["admin"] WHERE ANY(rr IN $auth.roles WHERE r = rr))
                AND
                    EXISTS(this.id) AND this.id = $thisOR0_auth_allow0_id OR ANY(r IN ["admin"] WHERE ANY(rr IN $auth.roles WHERE r = rr))
                `));
            expect(result[1]).toMatchObject({
                [`this_auth_allow0_id`]: sub,
                [`thisAND0_auth_allow0_id`]: sub,
                [`thisOR0_auth_allow0_id`]: sub,
            });
        });
    });
    describe("allow", () => {
        test("AND OR", () => {
            ["AND", "OR"].forEach((key) => {
                const idField = {
                    fieldName: "id",
                    typeMeta: {
                        name: "ID",
                        array: false,
                        required: false,
                        pretty: "String",
                        input: {
                            where: {
                                type: "String",
                                pretty: "String",
                            },
                            create: {
                                type: "String",
                                pretty: "String",
                            },
                            update: {
                                type: "String",
                                pretty: "String",
                            },
                        },
                    },
                    otherDirectives: [],
                    arguments: [],
                };
                // @ts-ignore
                const node = {
                    name: "Movie",
                    relationFields: [],
                    cypherFields: [],
                    enumFields: [],
                    scalarFields: [],
                    primitiveFields: [
                        idField,
                        {
                            fieldName: "title",
                            typeMeta: {
                                name: "String",
                                array: false,
                                required: false,
                                pretty: "String",
                                input: {
                                    where: {
                                        type: "String",
                                        pretty: "String",
                                    },
                                    create: {
                                        type: "String",
                                        pretty: "String",
                                    },
                                    update: {
                                        type: "String",
                                        pretty: "String",
                                    },
                                },
                            },
                            otherDirectives: [],
                            arguments: [],
                        },
                    ],
                    dateTimeFields: [],
                    interfaceFields: [],
                    objectFields: [],
                    pointFields: [],
                    authableFields: [idField],
                    auth: {
                        rules: [{ allow: { [key]: [{ id: "$jwt.sub" }, { id: "$jwt.sub" }, { id: "$jwt.sub" }] } }],
                        type: "JWT",
                    },
                };
                // @ts-ignore
                const neoSchema = {
                    nodes: [node],
                };
                const sub = randomstring_1.generate({
                    charset: "alphabetic",
                });
                // @ts-ignore
                const context = { neoSchema };
                context.jwt = {
                    sub,
                };
                const result = create_auth_and_params_1.default({
                    context,
                    entity: node,
                    allow: { parentNode: node, varName: "this" },
                });
                expect(utils_1.trimmer(result[0])).toEqual(utils_1.trimmer(`
                        (EXISTS(this.id) AND this.id = $this_auth_allow0_${key}0_id ${key} EXISTS(this.id) AND this.id = $this_auth_allow0_${key}1_id ${key} EXISTS(this.id) AND this.id = $this_auth_allow0_${key}2_id)
                    `));
                expect(result[1]).toMatchObject({
                    [`this_auth_allow0_${key}0_id`]: sub,
                    [`this_auth_allow0_${key}1_id`]: sub,
                    [`this_auth_allow0_${key}2_id`]: sub,
                });
            });
        });
    });
});
