// @neo4j/graphql/dist/translate/translate-delete.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("../constants");
const create_where_and_params_1 = __importDefault(require("./create-where-and-params"));
const create_auth_and_params_1 = __importDefault(require("./create-auth-and-params"));
const create_delete_and_params_1 = __importDefault(require("./create-delete-and-params"));
function translateDelete({ context, node }) {
    const { resolveTree } = context;
    const whereInput = resolveTree.args.where;
    const deleteInput = resolveTree.args.delete;
    const varName = "this";
    const matchStr = `MATCH (${varName}:${node.name})`;
    let whereStr = "";
    let allowStr = "";
    let deleteStr = "";
    let cypherParams = {};
    const whereStrs = [];
    if (whereInput) {
        const where = create_where_and_params_1.default({
            whereInput,
            varName,
            node,
            context,
            recursing: true,
        });
        if (where[0]) {
            whereStrs.push(where[0]);
            cypherParams = Object.assign(Object.assign({}, cypherParams), where[1]);
        }
    }
    const whereAuth = create_auth_and_params_1.default({
        operation: "DELETE",
        entity: node,
        context,
        where: { varName, node },
    });
    if (whereAuth[0]) {
        whereStrs.push(whereAuth[0]);
        cypherParams = Object.assign(Object.assign({}, cypherParams), whereAuth[1]);
    }
    if (whereStrs.length) {
        whereStr = `WHERE ${whereStrs.join(" AND ")}`;
    }
    const allowAuth = create_auth_and_params_1.default({
        operation: "DELETE",
        entity: node,
        context,
        allow: {
            parentNode: node,
            varName,
        },
    });
    if (allowAuth[0]) {
        cypherParams = Object.assign(Object.assign({}, cypherParams), allowAuth[1]);
        allowStr = `WITH ${varName}\nCALL apoc.util.validate(NOT(${allowAuth[0]}), "${constants_1.AUTH_FORBIDDEN_ERROR}", [0])`;
    }
    if (deleteInput) {
        const deleteAndParams = create_delete_and_params_1.default({
            context,
            node,
            deleteInput,
            varName,
            parentVar: varName,
            withVars: [varName],
        });
        [deleteStr] = deleteAndParams;
        cypherParams = Object.assign(Object.assign({}, cypherParams), deleteAndParams[1]);
    }
    const cypher = [matchStr, whereStr, deleteStr, allowStr, `DETACH DELETE ${varName}`];
    return [cypher.filter(Boolean).join("\n"), cypherParams];
}
exports.default = translateDelete;
