// grandstack/dist/commands/configure.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.handler = exports.builder = exports.desc = exports.command = void 0;

var _utils = require("../utils/");

const command = "configure";
exports.command = command;
const desc = "Setup configuration options for CLI";
exports.desc = desc;

const builder = yargs => {
  yargs.option("init", {
    alias: "i",
    description: "Initialize configuration",
    type: "boolean",
    default: false
  }).option("where-is", {
    alias: "w",
    description: "The current path to your configuration directory",
    type: "boolean",
    default: false
  }).option("profile", {
    alias: "p",
    description: "Create a new named profile",
    type: "string"
  }).option("neo4j-uri", {
    description: 'URI for the Neo4j instance. Example: "bolt://localhost:7798"',
    type: "string"
  }).option("neo4j-user", {
    description: "Database user",
    type: "string"
  }).option("neo4j-password", {
    description: "Database password for given user",
    type: "string"
  }).example(`$0 configure \\
    --neo4j-uri bolt://localhost:7687 \\
    --neo4j-user neo4j \\
    --neo4j-password letmein \\
    --new-profile graph-test`);
};

exports.builder = builder;

const handler = async ({
  profile,
  whereIs,
  init,
  "neo4j-uri": neo4j_uri,
  "neo4j-user": neo4j_user,
  "neo4j-password": neo4j_password
}) => {
  if (whereIs) {
    (0, _utils.infoExitZero)((0, _utils.whereIsConfigDir)());
  }

  if (init) {
    (0, _utils.createConfigDir)();
    (0, _utils.infoExitZero)((0, _utils.whereIsConfigDir)());
  }

  if (profile) {
    (0, _utils.checkCredentials)(neo4j_uri, neo4j_user, neo4j_password); // TODO: no creds run enquirer
    // How to handle low level versus interactive

    (0, _utils.createConfigDir)();
    const profilePath = (0, _utils.createProfileJson)({
      profile,
      neo4j_uri,
      neo4j_user,
      neo4j_password
    });
    (0, _utils.infoExitZero)(`Profile created at ${profilePath}`);
  }
};

exports.handler = handler;
