// @neo4j/graphql/dist/schema/point.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.cartesianPointDistance = exports.cartesianPointInput = exports.cartesianPoint = exports.pointDistance = exports.pointInput = exports.point = void 0;
exports.point = {
    name: "Point",
    fields: {
        longitude: {
            type: "Float!",
            resolve: (source) => {
                return source.point.x;
            },
        },
        latitude: {
            type: "Float!",
            resolve: (source) => {
                return source.point.y;
            },
        },
        height: {
            type: "Float",
            resolve: (source) => {
                return source.point.z;
            },
        },
        crs: "String!",
        srid: {
            type: "Int!",
            resolve: (source) => {
                return source.point.srid;
            },
        },
    },
};
exports.pointInput = {
    name: "PointInput",
    fields: {
        longitude: "Float!",
        latitude: "Float!",
        height: "Float",
    },
};
exports.pointDistance = {
    name: "PointDistance",
    fields: {
        point: "PointInput!",
        distance: {
            type: "Float!",
            description: "The distance in metres to be used when comparing two points",
        },
    },
};
exports.cartesianPoint = {
    name: "CartesianPoint",
    fields: {
        x: {
            type: "Float!",
            resolve: (source) => {
                return source.point.x;
            },
        },
        y: {
            type: "Float!",
            resolve: (source) => {
                return source.point.y;
            },
        },
        z: {
            type: "Float",
            resolve: (source) => {
                return source.point.z;
            },
        },
        crs: "String!",
        srid: {
            type: "Int!",
            resolve: (source) => {
                return source.point.srid;
            },
        },
    },
};
exports.cartesianPointInput = {
    name: "CartesianPointInput",
    fields: {
        x: "Float!",
        y: "Float!",
        z: "Float",
    },
};
exports.cartesianPointDistance = {
    name: "CartesianPointDistance",
    fields: {
        point: "CartesianPointInput!",
        distance: "Float!",
    },
};
